crates/sparrow-materialize/src/materialize/materialization_control.rs

```diff
@@ -0,0 +1,219 @@
+                    .change_context(Error::Internal)?
+                    .change_context(Error::Materialization)
+                {
+                    Ok(_) => {
+                        self.status = MaterializationProgress::Stopped(progress.clone());
+                    }
+                    Err(e) => {
+                        self.status = MaterializationProgress::Failed(progress.clone(), e);
+                    }
+                }
+            }
+            MaterializationProgress::Stopped(_) => {
+                tracing::info!("cannot stop; materialization already stopped")
+            }
+            MaterializationProgress::Failed(_, _) => {
+                tracing::info!("cannot stop; materialization already failed")
+            }
+        };
+
+        Ok(())
+    }
+
+    /// Gets the status of the materialization.
+    ///
+    /// Note that progress is reported periodically, and may not contain the exact
+    /// up-to-date numbers.
+    pub fn get_status(&self) -> MaterializationStatus {
+        let progress = self.progress_receiver.borrow();
+        match &*progress {
+            MaterializationProgress::Uninitialized => MaterializationStatus {
+                id: self.id.clone(),
+                progress: ProgressInformation::default(),
+                state: State::Uninitialized,
+                error: None,
+            },
+            MaterializationProgress::Running(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Running,
+                error: None,
+            },
+            MaterializationProgress::Stopped(progress) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Stopped,
+                error: None,
+            },
+            MaterializationProgress::Failed(progress, e) => MaterializationStatus {
+                id: self.id.clone(),
+                progress: progress.clone(),
+                state: State::Failed,
+                error: Some(e.to_string()),
+            },
+        }
+    }
+}
+
+/// Publish status of a materialization.
+#[derive(Debug, Default)]
+pub struct MaterializationStatus {
+    /// Unique identifier of the materialization
+    pub id: String,
+    /// Progress information
+    pub progress: ProgressInformation,
+    /// Current state of the materialization
+    pub state: State,
+    /// Failure reason
+    ///
+    /// Only populated if the materialization has failed
+    pub error: Option<String>,
+}
+
+/// Public state of a materialization.
+#[derive(Debug)]
+pub enum State {
+    Uninitialized,
+    Running,
+    Stopped,
+    Failed,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self::Uninitialized
+    }
+}
+
+/// Internal progress information for a materialization.
+#[derive(Debug)]
+enum MaterializationProgress {
```

Comment:
May be better to line this up with the status a bit further. This would let us eliminate some duplication of code from things like get status.

For instance:

```
struct MaterializationProgress {
  state: State,
  progress: ProgressInformation,
  error: Option<error_stack::Report<Error>>,
}
```